src/BitIntegers.jl

```diff
@@ -311,14 +311,40 @@ promote_rule(::Type{Float16}, ::Type{<:XBI}) = Float16
 (|)(x::T, y::T) where {T<:XBI} = or_int(x, y)
 xor(x::T, y::T) where {T<:XBI} = xor_int(x, y)
 
->>( x::UBS, y::UBU) = ashr_int(x, y)
->>( x::UBU, y::UBU) = lshr_int(x, y)
->>>(x::UBI, y::UBU) = lshr_int(x, y)
-<<( x::UBI, y::UBU) = shl_int(x, y)
-
->>( x::UBI, y::Int) = ifelse(0 <= y, x >> unsigned(y),  x << unsigned(-y))
-<<( x::UBI, y::Int) = ifelse(0 <= y, x << unsigned(y),  x >> unsigned(-y))
->>>(x::UBI, y::Int) = ifelse(0 <= y, x >>> unsigned(y), x << unsigned(-y))
+# LLVM bit shifting (lshr_int, ashr_int and shl_int) has performance issues for large integers
+# The following functions allows much faster code (10x speedup for 1024 bit integers)
+
+# Note: this should be a power of 2
```

Comment:
LLVM generates a correct code for shifting. It is slow in general because it emits code for several shift lengths and, instead of testing and jumping into the correct branch, it runs all branches and selects the correct result later. When the second argument is constant, the emitted code is specialized for that shift length (it eliminates useless branches). The same happens when the second argument is known by LLVM to be smaller than 64.

The only reason that `SHIFT_SPLIT_NBITS` should be a power of 2 is that the code splits `y` into lower (`y & mask`) and higher (`hi`) bits. I haven't tested in 32-bit machines, it could be better to set `SHIFT_SPLIT_NBITS = 32` in that case (I'm unsure, though).

I haven't found any instance in which the code generated is incorrect.